commit 6f7d2f4e236a082c32a4610f99892d8b04e0ef10
Author: Trustin Lee <trustin@gmail.com>
Date:   Tue Aug 2 07:47:36 2011 +0900

    NETTY-417 client channel still open after close and wait
    
    * Fixed a bug in SslHandler where a write attempt made after SSLEngine is closed result in unnotified future.  Fixed by respecting the case where SslEngineResult.Status == CLOSED

diff --git a/src/main/java/org/jboss/netty/handler/ssl/SslHandler.java b/src/main/java/org/jboss/netty/handler/ssl/SslHandler.java
index 028168b..666cb71 100644
--- a/src/main/java/org/jboss/netty/handler/ssl/SslHandler.java
+++ b/src/main/java/org/jboss/netty/handler/ssl/SslHandler.java
@@ -659,6 +659,11 @@ public class SslHandler extends FrameDecoder
                                     channel, future, msg, channel.getRemoteAddress());
                             offerEncryptedWriteRequest(encryptedWrite);
                             offered = true;
+                        } else if (result.getStatus() == Status.CLOSED) {
+                            // SSLEngine has been closed already.
+                            // Any further write attempts should be denied.
+                            success = false;
+                            break;
                         } else {
                             final HandshakeStatus handshakeStatus = result.getHandshakeStatus();
                             handleRenegotiation(handshakeStatus);
